Guard clause prevents `match` from seeing that all paths return

**Bug Report**

In a `match` statement in which all `case`s include a return statement, _mypy_ 0.942 falsely reports "Missing return statement [return]" if at least one `case` includes a guard clause.

**To Reproduce**

```
def func(e) -> int:
    match e:
        case x if True:
            return x
        case _:
            return 0
```

Or even more simply:
```
def func(e) -> int:
    match e:
        case _ if True:
            return 0
```

**Expected Behavior**

This code should type-check.

**Actual Behavior**

```
$ mypy --python-version 3.10 --show-error-codes x5.py
x5.py:1: error: Missing return statement  [return]
Found 1 error in 1 file (checked 1 source file)
```

**Notes**

If we change the first case to something that does not include a guard clause, like this:
```
def func(e) -> int:
    match e:
        case (x, _):
            return x
        case _:
            return 0
```
then _mypy_ returns no errors.

**Environment**

- Mypy version used: 0.942
- Mypy command-line flags:  --python-version 3.10 --show-error-codes
- Mypy configuration options from `mypy.ini` (and other config files): (none)
- Python version used: 3.10.2
- Operating system and version: Ubuntu 18.04

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -4944,7 +4944,7 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
                         self.push_type_map(pattern_map)
                         self.push_type_map(pattern_type.captures)
                     if g is not None:
-                        with self.binder.frame_context(can_skip=True, fall_through=3):
+                        with self.binder.frame_context(can_skip=False, fall_through=3):
                             gt = get_proper_type(self.expr_checker.accept(g))
 
                             if isinstance(gt, DeletedType):
@@ -4952,6 +4952,21 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
 
                             guard_map, guard_else_map = self.find_isinstance_check(g)
                             else_map = or_conditional_maps(else_map, guard_else_map)
+
+                            # If the guard narrowed the subject, copy the narrowed types over
+                            if isinstance(p, AsPattern):
+                                case_target = p.pattern or p.name
+                                if isinstance(case_target, NameExpr):
+                                    for type_map in (guard_map, else_map):
+                                        if not type_map:
+                                            continue
+                                        for expr in list(type_map):
+                                            if not (
+                                                isinstance(expr, NameExpr)
+                                                and expr.fullname == case_target.fullname
+                                            ):
+                                                continue
+                                            type_map[s.subject] = type_map[expr]
 
                             self.push_type_map(guard_map)
                             self.accept(b)

--- a/test-data/unit/check-python310.test
+++ b/test-data/unit/check-python310.test
@@ -1372,7 +1372,7 @@ match m:
         reveal_type(m)  # N: Revealed type is "__main__.Medal"
 
 [case testMatchNarrowUsingPatternGuardSpecialCase]
-def f(x: int | str) -> int:  # E: Missing return statement
+def f(x: int | str) -> int:
     match x:
         case x if isinstance(x, str):
             return 0
@@ -1972,4 +1972,47 @@ def f2(x: T) -> None:
             pass
         case DataFrame():  # type: ignore[misc]
             pass
+[builtins fixtures/primitives.pyi]
+
+[case testMatchGuardReachability]
+# flags: --warn-unreachable
+def f1(e: int) -> int:
+    match e:
+        case x if True:
+            return x
+        case _:
+            return 0  # E: Statement is unreachable
+    e = 0  # E: Statement is unreachable
+
+
+def f2(e: int) -> int:
+    match e:
+        case x if bool():
+            return x
+        case _:
+            return 0
+    e = 0  # E: Statement is unreachable
+
+def f3(e: int | str | bytes) -> int:
+    match e:
+        case x if isinstance(x, int):
+            return x
+        case [x]:
+            return 0  # E: Statement is unreachable
+        case str(x):
+            return 0
+    reveal_type(e)  # N: Revealed type is "builtins.bytes"
+    return 0
+
+def f4(e: int | str | bytes) -> int:
+    match e:
+        case int(x):
+            pass
+        case [x]:
+            return 0  # E: Statement is unreachable
+        case x if isinstance(x, str):
+            return 0
+    reveal_type(e)  # N: Revealed type is "Union[builtins.int, builtins.bytes]"
+    return 0
+
 [builtins fixtures/primitives.pyi]

